python/ql/src/Security/CWE-327/Ssl.qll

```diff
@@ -0,0 +1,186 @@
+    result.getNode() in [node.getArg(0), node.getArgByName("protocol")]
+  }
+}
+
+class SSLDefaultContextCreation extends ContextCreation {
+  SSLDefaultContextCreation() {
+    this = API::moduleImport("ssl").getMember("create_default_context").getACall()
+  }
+
+  // Allowed insecure versions are "TLSv1" and "TLSv1_1"
+  // see https://docs.python.org/3/library/ssl.html#context-creation
+  override DataFlow::CfgNode getProtocol() { none() }
+}
+
+class WrapSocketCall extends ConnectionCreation {
+  override CallNode node;
+
+  WrapSocketCall() { node.getFunction().(AttrNode).getName() = "wrap_socket" }
+
+  override DataFlow::CfgNode getContext() {
+    result.getNode() = node.getFunction().(AttrNode).getObject()
+  }
+}
+
+class OptionsAugOr extends ProtocolRestriction {
+  ProtocolVersion restriction;
+
+  OptionsAugOr() {
+    exists(AugAssign aa, AttrNode attr, Expr flag |
+      aa.getOperation().getOp() instanceof BitOr and
+      aa.getTarget() = attr.getNode() and
+      attr.getName() = "options" and
+      attr.getObject() = node and
+      flag = API::moduleImport("ssl").getMember("OP_NO_" + restriction).getAUse().asExpr() and
+      (
+        aa.getValue() = flag
+        or
+        impliesValue(aa.getValue(), flag, false, false)
+      )
+    )
+  }
+
+  override DataFlow::CfgNode getContext() { result = this }
+
+  override ProtocolVersion getRestriction() { result = restriction }
+}
+
+class OptionsAugAndNot extends ProtocolUnrestriction {
+  ProtocolVersion restriction;
+
+  OptionsAugAndNot() {
+    exists(AugAssign aa, AttrNode attr, Expr flag, UnaryExpr notFlag |
+      aa.getOperation().getOp() instanceof BitAnd and
+      aa.getTarget() = attr.getNode() and
+      attr.getName() = "options" and
+      attr.getObject() = node and
+      notFlag.getOp() instanceof Invert and
+      notFlag.getOperand() = flag and
+      flag = API::moduleImport("ssl").getMember("OP_NO_" + restriction).getAUse().asExpr() and
+      (
+        aa.getValue() = notFlag
+        or
+        impliesValue(aa.getValue(), notFlag, true, true)
+      )
+    )
+  }
+
+  override DataFlow::CfgNode getContext() { result = this }
+
+  override ProtocolVersion getUnrestriction() { result = restriction }
+}
+
+/** Whether `part` evaluates to `partIsTrue` if `whole` evaluates to `wholeIsTrue`. */
+predicate impliesValue(BinaryExpr whole, Expr part, boolean partIsTrue, boolean wholeIsTrue) {
+  whole.getOp() instanceof BitAnd and
+  (
+    wholeIsTrue = true and partIsTrue = true and part in [whole.getLeft(), whole.getRight()]
+    or
+    wholeIsTrue = true and
+    impliesValue([whole.getLeft(), whole.getRight()], part, partIsTrue, wholeIsTrue)
+  )
+  or
+  whole.getOp() instanceof BitOr and
+  (
+    wholeIsTrue = false and partIsTrue = false and part in [whole.getLeft(), whole.getRight()]
+    or
+    wholeIsTrue = false and
+    impliesValue([whole.getLeft(), whole.getRight()], part, partIsTrue, wholeIsTrue)
+  )
+}
```

Comment:
I would like some more extensive docs, with examples.

Reply:
Reasonable ask. We might also want to move this to a shared place at some point, and have a more comprehensive treatment of flags..